### PR TITLE
backupccl: create tree.SystemUsers, a new DescriptorCoverage enum

### PR DESCRIPTION
**Pull request by msbutler**
Previously during planning and execution RESTORE SYSTEM USERS was identified by
a `jobDetails` field. This refactor now identifies this flavor of
restore with a new  DescriptorCoverage enum value, `tree.SystemUsers.

This refactor eases the logic around exposing extra processing steps for
flavors of backup/restore that target different sets of descriptors.

Release note: None

### CODE REVIEW

File: pkg/ccl/backupccl/restore_job.go [hunk 1]
```diff
@@ -1265,6 +1265,12 @@ func (r *restoreResumer) doResume(ctx context.Context, execCtx interface{}) erro
 		return err
 	}
 
+	// To maintain mixed version compatibility where we plan RESTORE SYSTEM USERS on 22.1
+	// (not aware of the tree.SystemUsers enum) but execute on 22.2.
```
Comment (adityamaru) [inline pkg/ccl/backupccl/restore_job.go:1269]: nit: wrap comments at 70 characters

another nit: the comment sounds incomplete imo maybe reword to:

// A `RESTORE SYSTEM USERS` planned on a 22.1 node will use the `RestoreSystemUsers` field
// in the job details to identify this flavour of RESTORE. We must continue to check this field
// for mixed-version compatability.
//
// TODO(msbutler): Delete in 23.1

File: pkg/jobs/jobspb/jobs.proto [hunk 2]
```diff
@@ -329,6 +329,9 @@ message RestoreDetails {
 
   // DebugPauseOn describes the events that the job should pause itself on for debugging purposes.
   string debug_pause_on = 20;
+
+  // RestoreSystemUsers set to true if user runs RESTORE SYSTEM USERS.
+  // TODO (msbutler): deprecate after 22.2 branch cut
```
Comment (adityamaru) [inline pkg/jobs/jobspb/jobs.proto:334]: nit: is set to true

Also "Delete in 23.1" instead of deprecate maybe?

File: pkg/sql/sem/tree/backup.go [hunk 3]
```diff
@@ -28,10 +28,15 @@ const (
...
 	// by a `BACKUP TO` command.
 	RequestedDescriptors DescriptorCoverage = iota
+
 	// AllDescriptors table coverage means that backup is guaranteed to have all the
 	// relevant data in the cluster. These can only be created by running a
 	// full cluster backup with `BACKUP TO`.
 	AllDescriptors
+
+	// SystemUsers coverage indicates that only the system.users
+	// table will covered in the restore.
```
Comment (adityamaru) [inline pkg/sql/sem/tree/backup.go:38]: nit: only the `system.users` table will be restored from the backup.

File: pkg/sql/sem/tree/backup.go [hunk 4]
```diff
@@ -28,10 +28,15 @@ const (
 	// backup is not said to have complete table coverage unless it was created
```
Comment (adityamaru) [inline pkg/sql/sem/tree/backup.go:28]: I wonder whether we should change the comment above `type DescriptorCoverage` now that we're using this enum a little more liberally. Something like:

// DescriptorCoverage specifies the subset of descriptors that are requested during a backup
// or a restore.

File: pkg/ccl/backupccl/restore_planning.go [hunk 5]
```diff
@@ -1105,14 +1103,12 @@ func restorePlanHook(
 
 	var newDBNameFn func() (string, error)
 	if restoreStmt.Options.NewDBName != nil {
-		if restoreStmt.DescriptorCoverage == tree.AllDescriptors || len(restoreStmt.Targets.Databases) != 1 {
+		if restoreStmt.DescriptorCoverage == tree.AllDescriptors || len(restoreStmt.Targets.
```
Comment (adityamaru) [inline pkg/ccl/backupccl/restore_planning.go:1106]: nit: lets not break a newline on `.` maybe move the whole len clause to the new line. Also the error returned is specific to `RESTORE DATABASE` and multiple databases. I don't think it makes sense to return this error when using `new_db_name` with `RESTORE SYSTEM USERS`.

File: pkg/ccl/backupccl/restore_planning.go [hunk 6]
```diff
@@ -1834,7 +1837,7 @@ func doRestorePlan(
 			RevalidateIndexes:  revalidateIndexes,
 			DatabaseModifiers:  databaseModifiers,
 			DebugPauseOn:       debugPauseOn,
-			RestoreSystemUsers: restoreStmt.SystemUsers,
+			RestoreSystemUsers: restoreSystemUsers,
```
Comment (adityamaru) [inline pkg/ccl/backupccl/restore_planning.go:1840]: nit: maybe change to `RestoreSystemUsers: restoreStmt.DescriptorCoverage == tree.SystemUsers` and move your comment above the field?

File: pkg/ccl/backupccl/restore_job.go [hunk 7]
```diff
@@ -1509,7 +1515,7 @@ func (r *restoreResumer) doResume(ctx context.Context, execCtx interface{}) erro
 		if err := r.cleanupTempSystemTables(ctx, nil /* txn */); err != nil {
 			return err
 		}
-	} else if details.RestoreSystemUsers {
+	} else if details.DescriptorCoverage == tree.SystemUsers {
```
Comment (adityamaru) [inline pkg/ccl/backupccl/restore_job.go:1518]: It makes me a little uneasy thinking about what copy of details we're reading from here, and if we have indeed set and persisted `tree.SystemUsers` in a details that is coming from an older node. How do you feel about being more defensive and writing a helper method `isSystemUserRestore` which will check `details.DescriptorCoverage == tree.SystemUsers || details.RestoreSystemUsers`. We can them remove this helper and only check coverage in 23.1?

**Review by adityamaru — CHANGES_REQUESTED** [review]

File: pkg/ccl/backupccl/restore_job.go (hunk 1 above)
Comment (msbutler) [reply pkg/ccl/backupccl/restore_job.go]: Done.

File: pkg/ccl/backupccl/restore_job.go (hunk 7 above)
Comment (msbutler) [reply pkg/ccl/backupccl/restore_job.go]: very good call. Done.

File: pkg/ccl/backupccl/restore_planning.go (hunk 5 above)
Comment (msbutler) [reply pkg/ccl/backupccl/restore_planning.go]: Done.

File: pkg/ccl/backupccl/restore_planning.go (hunk 6 above)
Comment (msbutler) [reply pkg/ccl/backupccl/restore_planning.go]: Done.

File: pkg/jobs/jobspb/jobs.proto (hunk 2 above)
Comment (msbutler) [reply pkg/jobs/jobspb/jobs.proto]: Done.

File: pkg/sql/sem/tree/backup.go (hunk 4 above)
Comment (msbutler) [reply pkg/sql/sem/tree/backup.go:27]: Done.

File: pkg/sql/sem/tree/backup.go (hunk 3 above)
Comment (msbutler) [reply pkg/sql/sem/tree/backup.go]: Done.

**Review by adityamaru — APPROVED** [review]
